### DISCUSSION TITLE
Martensite - Packet

### DISCUSSION THREAD

**Discussion by ArianeNeves** [mtex-toolbox/mtex#1005] [Ask Anything]
Hello,

I'd like to determine the packet size of martensite. The job.transformedGrains.boundary store the parent and child boundaries. I tried to identify grains which have the same packetId or to merge the child grains with the misorientation of packet. But, I was unsuccessful.

Maybe someone can help me?

**Comment by nyyssont** [comment]
I suggest merging of neighbors based on similar packetId AND low misorientation angle between prior austenite orientations. Minimal example below using MTEX 5.7.0:

<details><summary><b>Click to view script: Packet merge</b></summary>
<p>

```matlab
% load the data
mtexdata martensite 
plotx2east

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize < 3)).rotations = quaternion.nan;
ebsd(grains(grains.grainSize < 3)).phase = 0;
ebsd(grains(grains.grainSize < 3)).grainId = 0;
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

%%
job = parentGrainReconstructor(ebsd,grains); % define job
job.p2c = orientation.KurdjumovSachs(job.p2c.CS,job.p2c.SS);
job.calcParent2Child; % get representative OR

%%
job.calcGBVotes('numFit',4,'count','threshold',3*degree) % compute votes from child grain boundaries
job.calcParentFromVote('minProb',4)

%%
for l = 1:2
    job.calcGBVotes('p2c','threshold',5*degree) % compute votes from p2c grain boundaries
    job.calcParentFromVote
end

%% Packet merge
job.calcVariants

% Packet boundaries merge on two conditions: same packetId and low
% misorientation between parent orientations

z = job.packetId
zz = job.transformedGrains.neighbors
neigh_pIds = z(zz);
misos = inv(job.grains(zz(:,1)).meanOrientation).*job.grains(zz(:,2)).meanOrientation;
conditions = neigh_pIds(:,1) == neigh_pIds(:,2) & misos.angle<5*degree;
mergeIds = zz(conditions,:);

packets = merge(job.grains,mergeIds)

plot(job.grains,job.packetId)
hold on
plot(packets.boundary,'linewidth',2,'linecolor','magenta')

figure
histogram(packets('Iron fcc').equivalentRadius*2)
title('Packet size, equivalent circle diameter')
```

</p>
</details>

The script should give the following outputs:

![newfile](https://user-images.githubusercontent.com/10060198/117100945-5db73a00-ad75-11eb-83dd-1891e0e8fd9e.jpg)
![ECD](https://user-images.githubusercontent.com/10060198/117100954-61e35780-ad75-11eb-919a-7d8034eafbc7.jpg)

**Reply by ArianeNeves** [reply]
Hello Tuomo!
Thank you for your help!
I have been tried to use your code in my data. I realized that you use job.grains which have Iron (Alpha) and Iron (Gamma). It can be observed that the code also identified packet boundaries inside the non-reconstructed areas. 

I think it would be better to use z = job.transformedGrains.packetId. But I can not determine the neighbors packets. I am confusing in the second and third lines. May you can help me?

z = job.transformedGrains.packetId.
zz = job.transformedGrains.neighbors
neigh_pIds = z(zz);

misos = inv(job.grains(('Iron (Gamma)'),(zz(:,1))).meanOrientation).*job.grains(('Iron (Gamma)'),(zz(:,2))).meanOrientation;
conditions = neigh_pIds(:,1) == neigh_pIds(:,2) & misos.angle<5*degree;
mergeIds = zz(conditions,:);

packets = merge(job.grains('Iron (Gamma)'),mergeIds)

plot(job.grains('Iron (Gamma)'),job.transformedGrains.packetId)
hold on
plot(packets.boundary,'linewidth',2,'linecolor','magenta')

Thank you!

**Comment by PietroMart** [comment]
Hi!

I tried to apply the code you proposed, but I am getting this error:

Error using subsind (line 68)
Out of range; index must be a positive integer or logical.

Error in grain2d/subsref (line 25)
    ind = subsind(grains,s(1).subs);

Error in parent_grain_reconstructor (line 225)
misos = inv(job.grains(zz(:,1)).meanOrientation).*job.grains(zz(:,2)).meanOrientation;

any idea what could be the problem?
